smartsim/_core/_cli/build.py

```diff
@@ -331,84 +328,146 @@ def infer_torch_device(self) -> str:
-            if (inferred_device == "gpu") and (device == "cpu"):
-                logger.warning("CPU requested, but GPU backend is available")
-            _install_torch_from_pip(self.versions, device, self.verbose)
-        # if torch already installed, check the versions to make sure correct
-        # torch version is downloaded for that particular device
-        else:
-            installed = Version_(pkg_resources.get_distribution("torch").version)
-            if device == "gpu":
-                # if torch version is x.x.x+cpu
-                if "cpu" in installed.patch:
-                    msg = (
-                        "Torch CPU is currently installed but torch GPU requested. Uninstall all torch packages "
-                        + "and run the `smart build` command again to obtain Torch GPU libraries"
-                    )
-                    logger.warning(msg)
-
+        logger.info(f"Searching for a compatible TORCH install...")
+        if self.build_env.is_macos():
+            end_point = None
+            device_suffix = ""
+        else:  # linux
+            end_point = "https://download.pytorch.org/whl/torch_stable.html"
             if device == "cpu":
-                # if torch version if x.x.x then we need to install the cpu version
-                if "cpu" not in installed.patch and not self.build_env.is_macos():
-                    msg = (
-                        "Torch GPU installed in python environment but requested Torch CPU. "
-                        + " Run `pip uninstall torch torchvision` and run `smart build` again"
-                    )
-                    logger.error(msg)  # error because this is usually fatal
-            logger.info(f"Torch {self.versions.TORCH} installed in Python environment")
+                device_suffix = self.versions.TORCH_CPU_SUFFIX
+            elif device in ["gpu", "cuda"]:
+                device_suffix = self.versions.TORCH_CUDA_SUFFIX
+            else:
+                raise BuildError("Unrecognized device requested")
+
+        torch_packages = {
+            "torch": f"{self.versions.TORCH}{device_suffix}",
+            "torchvision": f"{self.versions.TORCHVISION}{device_suffix}",
+        }
+
+        def torch_validator(package: str, version: t.Optional[str]) -> bool:
+            if not self.build_env.check_installed(package, version):
+                return False
+            # Previous check only looks at major/minor version numbers,
+            # Torch requires we look at the patch as well
+            installed = self.build_env.get_package_version(package)
+            if device_suffix and device_suffix not in installed.patch:
+                # Torch requires that we compare to the patch
+                raise VersionConflictError(
+                    f"{package}=={installed} does not satisfy device "
+                    f"suffix requirement: {device_suffix}"
+                )
+            return True
```

Comment:
Nesting functions inside functions can make code hard to read. Are you doing this just to limit access scope? Might be better to take it out and just throw a few underscores at its name

Reply:
Good point! This was just to limit scope, but probably not worth the cost of readability. Moved to a "private" static method on the Build class!

Reply:
On a second pass, the reason for this was also in part because I was catching the `device_suffix` var in the closure. I broke out the construction of the torch version validation function into a separate factory method which is hopefully a bit more readable.

Let me know if you think this is now sufficiently clear, otherwise I can take another pass to try and clean it up in a more substantial way!